### ISSUE TITLE
Silent data loss when response contains characters invalid for XML PCDATA

### ISSUE THREAD

**Issue by jnm** [enketo/enketo-express#181]
Hi @MartijnR, some people are copying text from PDFs that contains old ASCII control characters and pasting it as responses to questions. I'm guessing that, depending on their platform, these characters might be invisible to them.

The problem is that these characters break submissions and draft saving/loading, but there's not an indication of an error until it's too late to retrieve the responses already entered.

### Form used to test (single text question)
```xml
<?xml version="1.0" encoding="utf-8"?>
<h:html xmlns="http://www.w3.org/2002/xforms" xmlns:ev="http://www.w3.org/2001/xml-events" xmlns:h="http://www.w3.org/1999/xhtml" xmlns:jr="http://openrosa.org/javarosa" xmlns:odk="http://www.opendatakit.org/xforms" xmlns:orx="http://openrosa.org/xforms" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
  <h:head>
    <h:title>text</h:title>
    <model>
      <instance>
        <text id="text">
          <formhub>
            <uuid/>
          </formhub>
          <text/>
          <meta>
            <instanceID/>
          </meta>
        </text>
      </instance>
      <bind nodeset="/text/text" type="string"/>
      <bind calculate="concat('uuid:', uuid())" nodeset="/text/meta/instanceID" readonly="true()" type="string"/>
      <bind calculate="'dbfdf102b9e74a69abdcc8527a702275'" nodeset="/text/formhub/uuid" type="string"/>
    </model>
  </h:head>
  <h:body>
    <input ref="/text/text">
      <label>text</label>
    </input>
  </h:body>
</h:html>
```

### Reproduce via drafts

1. Open the form;
1. Paste something funky as the text response, e.g. `whatthehey`;
    * :information_source: It looks like GitHub is not preserving the characters. Perhaps generate the text response with `console.log('what' + String.fromCharCode(4) + 'the' + String.fromCharCode(28) + 'hey')`;
1. Save the submission as a draft;
1. Notice no errors in the UI or developer console;
1. Attempt to load the draft;
1. See "Loading Error…Error trying to parse XML record. Invalid XML source".

### Reproduce via submission

1. Use Google Chrome;
1. Create funky text response as above;
1. Submit;
1. See `parseXML error` in the console;
1. See `text - 1 was successfully submitted` in the UI;
1. Notice (using the browser's developer tools) that the `POST`ed form data contains a valid root node and form UUID:
    ```
    ------WebKitFormBoundaryBGt3di99CAJQp9hF
    Content-Disposition: form-data; name="xml_submission_file"; filename="xml_submission_file"
    Content-Type: text/xml

    <text xmlns:jr="http://openrosa.org/javarosa" xmlns:odk="http://www.opendatakit.org/xforms" xmlns:orx="http://openrosa.org/xforms" id="text"><parsererror xmlns="http://www.w3.org/1999/xhtml" style="display: block; white-space: pre; border: 2px solid #c77; padding: 0 1em 0 1em; margin: 1em; background-color: #fdd; color: black"><h3>This page contains the following errors:</h3><div style="font-family:monospace;font-size:12px">error on line 5 at column 21: PCDATA invalid Char value 4</div><h3>Below is a rendering of the page up to the first error.</h3></parsererror>
              <formhub>
                <uuid>dbfdf102b9e74a69abdcc8527a702275</uuid>
              </formhub>
              <text/></text>
    ------WebKitFormBoundaryBGt3di99CAJQp9hF--
    ```
1. Notice, at least in the case of KoBoCAT, that this incomplete XML is indeed accepted as a valid submission, and a 201 response is returned to Enketo;
1. Observe that Enketo has deleted the submission from browser storage, as is typical after a successful upload.

**Comment by MartijnR** [comment]
Thanks. What do you think the best solution is? Some ideas:

Should we perhaps just strip these characters before storing text (type=string binds) into Enketo's model? 

Or should we prevent pasting such text (like we do with mask in number inputs)?

I'm guessing the latter will leave users very puzzled about why it won't allow these characters with no idea on how to resolve it (even with a helpful error message).

**Comment by tinok** [comment]
I would vote option 1 definitely. Just strip these characters, no need to
preserve them.

**Comment by MartijnR** [comment]
Makes sense. If KoBo could take this on, I'd welcome a PR with a test(s) and I'll review. 

Assuming these characters don't cause any issues before then, the location for the fix would likely be in here: https://github.com/enketo/enketo-core/blob/master/src/js/form-model.js#L1444

**Comment by MartijnR** [comment]
A third option would be:

* detect paste event and remove those characters before letting past proceed. This will remove them from the input element itself. 

I think this is less preferable than doing it in the model though, because avoiding new event handlers will lead to fewer bugs. So, it's a backup solution if these characters also cause problems in HTML (in which case [mask.js](https://github.com/enketo/enketo-core/blob/master/src/js/mask.js) probably has useful code to reuse).

**Comment by tinok** [comment]
Update: I tested it with a source PDF and was able to reproduce the issue without needing to generate invalid chars in the console. I copied the text into this file: [invalid characters.txt](https://github.com/enketo/enketo-express/files/4827313/invalid.characters.txt)

The issue stems from the fact that ligature letters (like the combined ff or fi characters) are copied as these ASCII control characters. Here is the same abstract as in the text file with the ligatures highlighted: 
![image](https://user-images.githubusercontent.com/673511/85615036-9cdf0f80-b629-11ea-9a72-27176bf80304.png)

**Comment by jnm** [comment]
Stripping these characters out seems like an alright real-world solution and may very well save other people pain down the line—but it isn't really Enketo's _job_, and I don't get a warm, fuzzy feeling about silently modifying the inputted responses. Can we escape the characters in the same way `<`, `>`, and friends are handled?

**Comment by tinok** [comment]
Do you mean replace them with a different valid character, like � (U+FFFD, for replacements)?

I would argue that we can safely strip them because 1) these are never text that a user can enter, only unintentional characters that can't be used for data analysis, 2) EE already strips similar things like rich text formatting or images that may have been copied, 3) they are being stripped by similar user-level applications such as MS Word. 

Isn't it possible that preserving them could lead to more difficulties down the road

**Comment by tinok** [comment]
> EE already strips similar things like rich text formatting or images that may have been copied
>> The browser does that, not EE.

That's a good point. Would be great if the browser stripped the control characters out as well when pasting. I guess we've documented now why we have to strip them in EE? 🤷

**Comment by gido3823** [comment]
سوف اذهب شكرا
